lib/block_view_validator.go

```diff
@@ -0,0 +1,1415 @@
+	currentValidatorEntry := &ValidatorEntry{
+		ValidatorID:                validatorID,
+		ValidatorPKID:              transactorPKIDEntry.PKID,
+		Domains:                    txMeta.Domains,
+		DisableDelegatedStake:      txMeta.DisableDelegatedStake,
+		VotingPublicKey:            txMeta.VotingPublicKey,
+		VotingPublicKeySignature:   txMeta.VotingPublicKeySignature,
+		VotingSignatureBlockHeight: txMeta.VotingSignatureBlockHeight,
+		TotalStakeAmountNanos:      totalStakeAmountNanos,
+		RegisteredAtBlockHeight:    registeredAtBlockHeight,
+		ExtraData:                  mergeExtraData(prevExtraData, txn.ExtraData),
+	}
+	// Set the ValidatorEntry.
+	bav._setValidatorEntryMappings(currentValidatorEntry)
+
+	// Add a UTXO operation
+	utxoOpsForTxn = append(utxoOpsForTxn, &UtxoOperation{
+		Type:               OperationTypeRegisterAsValidator,
+		PrevValidatorEntry: prevValidatorEntry,
+		// PrevStakeEntries: prevStakeEntries, // TODO: in subsequent PR
+	})
+	return totalInput, totalOutput, utxoOpsForTxn, nil
+}
+
+func (bav *UtxoView) _disconnectRegisterAsValidator(
+	operationType OperationType,
+	currentTxn *MsgDeSoTxn,
+	txHash *BlockHash,
+	utxoOpsForTxn []*UtxoOperation,
+	blockHeight uint32,
+) error {
+	// Validate the last operation is a RegisterAsValidator operation.
+	if len(utxoOpsForTxn) == 0 {
+		return fmt.Errorf("_disconnectRegisterAsValidator: utxoOperations are missing")
+	}
+	operationIndex := len(utxoOpsForTxn) - 1
+	operationData := utxoOpsForTxn[operationIndex]
+	if operationData.Type != OperationTypeRegisterAsValidator {
+		return fmt.Errorf(
+			"_disconnectRegisterAsValidator: trying to revert %v but found %v",
+			OperationTypeRegisterAsValidator,
+			operationData.Type,
+		)
+	}
+
+	// Convert TransactorPublicKey to TransactorPKID.
+	transactorPKIDEntry := bav.GetPKIDForPublicKey(currentTxn.PublicKey)
+	if transactorPKIDEntry == nil || transactorPKIDEntry.isDeleted {
+		return RuleErrorInvalidValidatorPKID
+	}
+
+	// Delete the current ValidatorEntry.
+	currentValidatorEntry, err := bav.GetValidatorByPKID(transactorPKIDEntry.PKID)
+	if err != nil {
+		return errors.Wrapf(err, "_disconnectRegisterAsValidator: ")
+	}
+	if currentValidatorEntry == nil {
+		return fmt.Errorf(
+			"_disconnectRegisterAsValidator: no ValidatorEntry found for %v", transactorPKIDEntry.PKID,
+		)
+	}
+	bav._deleteValidatorEntryMappings(currentValidatorEntry)
+
+	// Restore the PrevValidatorEntry, if exists. The PrevValidatorEntry won't exist if this was
+	// the first time this ValidatorEntry was created. The PrevValidatorEntry will exist if this
+	// was an update operation on an existing ValidatorEntry.
+	prevValidatorEntry := operationData.PrevValidatorEntry
+	if prevValidatorEntry != nil {
+		bav._setValidatorEntryMappings(prevValidatorEntry)
+	}
+
+	// TODO: In subsequent PR, if PrevStakeEntries, delete the
+	// current StakeEntries and restore the prev StakeEntries.
+	// This should also update GlobalStakeAmountNanos.
```

Comment:
In a subsequent PR, we do track PrevGlobalStakeAmountNanos on the UtxoOperation exactly as you suggest here so that it is easier to revert. For now, I am just going to leave this TODO.